In [29]:
import pandas as pd


In [27]:
df = pd.read_csv('../data/raw/phdb_dinosaur_occurrences.csv')
df.head()

C:\Users\john_\AppData\Local\Temp\ipykernel_23072\3316100536.py:1: DtypeWarning: Columns (56,81,82,83,84,119,131) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../data/raw/phdb_dinosaur_occurrences.csv')


,occurrence_no,record_type,collection_no,identified_name,identified_rank,accepted_name,accepted_attr,accepted_rank,accepted_no,early_interval,...,max_ma_error,rare_body_parts,min_ma_error,direct_ma_value,direct_ma_unit,direct_ma_method,zone_type,direct_ma_error,plant_organ,artifacts
0,41524,occ,3257,Aves indet.,class,Aves,(Linnaeus 1758),class,36616,Lutetian,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,41580,occ,3256,Aves indet.,class,Aves,(Linnaeus 1758),class,36616,Ypresian,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,130209,occ,10755,Chaoyangosaurus liaosiensis n. gen. n. sp.,species,Chaoyangsaurus youngi,Zhao et al. 1999,species,55580,Late Kimmeridgian,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,130294,occ,10764,Protarchaeopteryx robusta n. gen. n. sp.,species,Protarchaeopteryx robusta,Ji and Ji 1997,species,66068,Late Barremian,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,130295,occ,10764,Caudipteryx zoui n. gen. n. sp.,species,Caudipteryx zoui,Ji et al. 1998,species,66066,Late Barremian,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [21]:
df.shape

(37817, 133)

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37817 entries, 0 to 37816
Columns: 133 entries, occurrence_no to artifacts
dtypes: float64(17), int64(4), object(112)
memory usage: 38.4+ MB


In [28]:
df.columns.tolist()

['occurrence_no',
 'record_type',
 'collection_no',
 'identified_name',
 'identified_rank',
 'accepted_name',
 'accepted_attr',
 'accepted_rank',
 'accepted_no',
 'early_interval',
 'max_ma',
 'min_ma',
 'ref_author',
 'ref_pubyr',
 'reference_no',
 'phylum',
 'class',
 'lng',
 'lat',
 'occurrence_comments',
 'collection_name',
 'cc',
 'state',
 'county',
 'latlng_basis',
 'latlng_precision',
 'geogscale',
 'geogcomments',
 'paleomodel',
 'geoplate',
 'paleoage',
 'paleolng',
 'paleolat',
 'formation',
 'geological_group',
 'stratscale',
 'stratcomments',
 'lithology1',
 'environment',
 'size_classes',
 'pres_mode',
 'collection_type',
 'research_group',
 'taxon_environment',
 'motility',
 'life_habit',
 'diet',
 'reproduction',
 'ontogeny',
 'composition',
 'architecture',
 'identified_no',
 'late_interval',
 'order',
 'family',
 'genus',
 'abund_value',
 'abund_unit',
 'collection_aka',
 'altitude_value',
 'altitude_unit',
 'lithdescript',
 'lithadj1',
 'fossilsfrom1',
 'preservation

In [23]:
column_profile = pd.DataFrame({
    "column": df.columns,
    "dtype": df.dtypes.astype(str),
    "missing": df.isna().sum(),
    "missing_pct": (df.isna().mean() * 100).round(2),
    "unique": df.nunique()
})

column_profile

,column,dtype,missing,missing_pct,unique
occurrence_no,occurrence_no,int64,0,0.00,37817
record_type,record_type,object,0,0.00,1
collection_no,collection_no,int64,0,0.00,14351
identified_name,identified_name,object,0,0.00,10505
identified_rank,identified_rank,object,0,0.00,15
...,...,...,...,...,...
direct_ma_method,direct_ma_method,object,36837,97.41,11
zone_type,zone_type,object,36605,96.80,8
direct_ma_error,direct_ma_error,float64,36930,97.65,76
plant_organ,plant_organ,object,37809,99.98,2


In [24]:
column_profile.sort_values("missing_pct", ascending=False).head(30)

,column,dtype,missing,missing_pct,unique
plant_organ,plant_organ,object,37809,99.98,2
regionalbedunit,regionalbedunit,object,37746,99.81,1
regionalorder,regionalorder,object,37665,99.60,2
regionalbed,regionalbed,object,37594,99.41,73
ecospace_comments,ecospace_comments,object,37590,99.40,7
regionalsection,regionalsection,object,37564,99.33,30
container_no,container_no,float64,37464,99.07,30
rare_body_parts,rare_body_parts,object,37324,98.70,34
lithification2,lithification2,object,37304,98.64,3
vision,vision,object,37212,98.40,1


In [25]:
column_profile.sort_values("missing_pct").head(30)

,column,dtype,missing,missing_pct,unique
occurrence_no,occurrence_no,int64,0,0.00,37817
record_type,record_type,object,0,0.00,1
collection_no,collection_no,int64,0,0.00,14351
identified_name,identified_name,object,0,0.00,10505
identified_rank,identified_rank,object,0,0.00,15
accepted_name,accepted_name,object,0,0.00,6172
accepted_rank,accepted_rank,object,0,0.00,15
accepted_no,accepted_no,int64,0,0.00,6175
ref_author,ref_author,object,1,0.00,3818
early_interval,early_interval,object,0,0.00,253


In [26]:
column_profile[
    (column_profile["missing_pct"] >= 20) &
    (column_profile["missing_pct"] <= 80)
].sort_values("missing_pct")

,column,dtype,missing,missing_pct,unique
geogcomments,geogcomments,object,7883,20.85,9925
stratscale,stratscale,object,8381,22.16,5
family,family,object,8733,23.09,409
paleolat,paleolat,float64,11704,30.95,4801
paleolng,paleolng,float64,11704,30.95,6009
collection_methods,collection_methods,object,12016,31.77,389
genus,genus,object,14244,37.67,3087
formation,formation,object,14328,37.89,1778
stratcomments,stratcomments,object,15714,41.55,5746
abund_unit,abund_unit,object,15919,42.09,8


#Initial Data Quality Observations

23 columns of 131 are 100% filled with severaly columns missing 50% or more entires. The entirty of the bottom 30 of percentages of completed data are above 90% meaning potentially optional columns not filled for almost all fossil occurences. 

In [34]:
df['occurrence_no'].duplicated().sum()

np.int64(0)

In [35]:
df['accepted_name'].nunique()

6172

In [37]:
df["accepted_rank"].value_counts()

accepted_rank
species           17316
unranked clade     6760
genus              6108
family             4554
suborder            738
subfamily           715
class               632
order               309
infraorder          238
superfamily         196
subspecies          147
subclass             62
tribe                30
superorder           10
subgenus              2
Name: count, dtype: int64

In [38]:
df["accepted_name"].value_counts().head(25)

accepted_name
Theropoda             1992
Dinosauria            1422
Sauropoda             1042
Hadrosauridae         1022
Aves                   632
Ornithopoda            521
Ceratopsidae           287
Dromaeosauridae        265
Titanosauria           248
Tyrannosauridae        245
Ankylosauria           240
Ornithischia           199
Grallator              195
Passeriformes          167
Coelurosauria          164
Ornithomimidae         159
Titanosauridae         156
Ankylosauridae         149
Ceratopsia             148
Abelisauridae          139
Carnosauria            136
Paronychodon           136
Eubrontes              132
Pinguinus impennis     126
Iguanodontidae         123
Name: count, dtype: int64

In [39]:
df["genus"].value_counts().head(25)

genus
Grallator            431
Anas                 336
Eubrontes            304
Phalacrocorax        278
Corvus               270
Puffinus             244
Richardoestesia      239
Falco                198
Megaloolithus        183
Turdus               177
Pterodroma           174
Buteo                170
Triceratops          166
Anomoepus            159
Larus                155
Saurornitholestes    148
Allosaurus           147
Paronychodon         142
Tyto                 139
Sula                 138
Camarasaurus         138
Iguanodon            133
Gallirallus          132
Branta               132
Pinguinus            127
Name: count, dtype: int64

In [40]:
df["family"].value_counts().head(25)

family
NO_FAMILY_SPECIFIED    3047
Anatidae               1731
Hadrosauridae          1625
Dromaeosauridae        1018
Grallatoridae           835
Accipitridae            760
Procellariidae          756
Ceratopsidae            747
Columbidae              744
Rallidae                727
Phasianidae             662
Alcidae                 545
Tyrannosauridae         512
Scolopacidae            507
Laridae                 493
Corvidae                492
Strigidae               431
Phalacrocoracidae       327
Ankylosauridae          293
Nodosauridae            273
Diplodocidae            270
Falconidae              260
Ornithomimidae          259
Elongatoolithidae       257
Sulidae                 251
Name: count, dtype: int64

In [41]:
df[["identified_name", "accepted_name", "identified_rank", "accepted_rank"]].head(20)

,identified_name,accepted_name,identified_rank,accepted_rank
0,Aves indet.,Aves,class,class
1,Aves indet.,Aves,class,class
2,Chaoyangosaurus liaosiensis n. gen. n. sp.,Chaoyangsaurus youngi,species,species
3,Protarchaeopteryx robusta n. gen. n. sp.,Protarchaeopteryx robusta,species,species
4,Caudipteryx zoui n. gen. n. sp.,Caudipteryx zoui,species,species
5,Theropoda indet.,Theropoda,unranked clade,unranked clade
6,Dinosauria indet.,Dinosauria,unranked clade,unranked clade
7,Gorgosaurus sternbergi n. sp.,Gorgosaurus libratus,species,species
8,Hadrosauridae indet.,Hadrosauridae,family,family
9,Gorgosaurus libratus,Gorgosaurus libratus,species,species


Checking taxonomy data on fossil occurances. Discovered that dataset does not only include Dinosaur fossils, but also decedents of dinosaurs. Will remove non-dinosaur fossils in cleaning stage to keep scope of project clean.

In [ ]:
#Start of timeline investigation
df["early_interval"].value_counts().head(30)

early_interval
Holocene               5409
Late Pleistocene       4772
Late Campanian         2617
Late Maastrichtian     1518
Maastrichtian          1386
Kimmeridgian           1110
Aptian                  958
Middle Campanian        957
Hettangian              800
Early Maastrichtian     702
Campanian               635
Lancian                 589
Late Kimmeridgian       560
Middle Pleistocene      541
Norian                  522
Late Barremian          490
Pleistocene             486
Early Cenomanian        471
Rancholabrean           437
Tithonian               436
Berriasian              406
Albian                  398
Cenomanian              362
Barremian               344
Early Campanian         344
Valanginian             314
Calabrian               312
Late Albian             263
Santonian               260
Early Barremian         252
Name: count, dtype: int64

In [43]:
df["late_interval"].value_counts().head(30)

late_interval
Tithonian              921
Holocene               807
Late Campanian         589
Sinemurian             553
Early Maastrichtian    518
Albian                 513
Early Cenomanian       263
Santonian              262
Maastrichtian          242
Early Aptian           219
Barremian              217
Late Turonian          198
Early Tithonian        170
Valanginian            168
Aptian                 161
Hauterivian            145
Early Albian           133
Rhaetian               122
Early Barremian        117
Middle Campanian       114
Pliensbachian          106
Early Campanian         98
Toarcian                92
Cenomanian              90
Turonian                83
Early Turonian          74
Late Pleistocene        69
Berriasian              67
Middle Cenomanian       61
Pleistocene             53
Name: count, dtype: int64

In [44]:
df["max_ma"].describe()

count    37817.000000
mean        71.029718
std         65.813394
min          0.011700
25%          0.129000
50%         72.200000
75%        121.400000
max        251.902000
Name: max_ma, dtype: float64

In [45]:
df["min_ma"].describe()

count    37817.000000
mean        64.545258
std         61.613984
min          0.000000
25%          0.011700
50%         66.000000
75%        100.500000
max        248.600000
Name: min_ma, dtype: float64

In [46]:
df[["max_ma", "min_ma"]].isna().sum()

max_ma    0
min_ma    0
dtype: int64

In [47]:
df[["early_interval", "late_interval", "max_ma", "min_ma"]].head(20)

,early_interval,late_interval,max_ma,min_ma
0,Lutetian,NaN,48.07,41.03
1,Ypresian,NaN,56.00,48.07
2,Late Kimmeridgian,Valanginian,152.21,132.60
3,Late Barremian,Early Aptian,125.77,119.57
4,Late Barremian,Early Aptian,125.77,119.57
5,Middle Campanian,Late Campanian,83.60,72.20
6,Berriasian,Aptian,143.10,113.20
7,Late Campanian,NaN,83.60,72.20
8,Late Campanian,NaN,83.60,72.20
9,Late Campanian,NaN,83.60,72.20


Timeline data suggests that fossils cover from 250 ma to 0 ma. 

In [48]:
#Start of Geography investigation
df['cc'].value_counts().head(30)

cc
US    11384
CA     2226
ES     2224
CN     1979
NZ     1268
AR     1214
UK     1184
FR     1037
MN      980
AU      831
IT      681
BR      585
MX      584
PT      533
MG      419
ZA      415
GR      395
RU      385
JP      356
MA      355
DE      336
RO      329
UZ      316
KR      312
TZ      307
IN      293
PF      280
BS      211
TO      211
KZ      204
Name: count, dtype: int64

In [50]:
df['state'].value_counts().head(30)

state
California         2227
Alberta            1679
Wyoming            1267
Montana            1016
New Mexico          968
Florida             908
Utah                901
England             832
Omnogov             685
Colorado            652
South Island        602
Texas               521
Aragón              389
La Rioja            334
Cataluña            333
Massachusetts       320
Virginia            283
Liaoning            280
Crete               270
North Dakota        258
Río Negro           253
Navoi               252
Neuquén             239
Arizona             230
Lindi               223
Coahuila            222
South Australia     214
Sichuan             213
South Dakota        208
Centro              201
Name: count, dtype: int64

In [51]:
df[['lat','lng']].describe()

,lat,lng
count,37817.000000,37817.000000
mean,24.991546,-19.977060
std,29.819721,92.787027
min,-84.333336,-179.154999
25%,18.330000,-106.624100
50%,37.404999,-17.209999
75%,43.575001,35.373600
max,89.039169,178.677002


In [52]:
df[['lat','lng']].isna().sum()

lat    0
lng    0
dtype: int64

In [53]:
df['latlng_basis'].value_counts()

latlng_basis
based on nearby landmark    12868
estimated from map          11994
stated in text               9005
based on political unit      1374
unpublished field data        963
Name: count, dtype: int64

In [54]:
df['latlng_precision'].value_counts()

latlng_precision
seconds    11999
6           6749
4           5610
3           4022
minutes     3608
2           1918
5           1830
1           1537
7            221
degrees      204
8            119
Name: count, dtype: int64

Geographic investigation shows countries listed by 2 letter identifiers. States are not only US states, but other country states included on list and are fully spelled out. No latitude or longitude data missing. 

In [55]:
#Start of Geology/Enviroment investigation
df['formation'].nunique()

1778

In [56]:
df['formation'].value_counts().head(30)

formation
Morrison                1164
Hell Creek               767
Dinosaur Park            727
Oldman                   339
Nemegt                   319
Lance                    318
Portland                 301
Aguja                    276
Elliot                   272
Fruitland/Kirtland       258
Cloverly                 226
Tendaguru                223
Kirtland                 222
Cedar Mountain           209
Two Medicine             195
Horseshoe Canyon         194
Djadokhta                179
Bissekty                 175
Judith River             175
Lameta                   168
San Diego                161
Baynshire                154
Conques                  150
Cerro del Pueblo         148
Allen                    141
Baruungoyot              128
Lourinhã                 127
Yixian                   126
Villar del Arzobispo     123
Ferris                   121
Name: count, dtype: int64

In [72]:
df['environment'].nunique()

70

In [57]:
df['environment'].value_counts()

environment
terrestrial indet.            14527
cave                           4589
marine indet.                  1709
fluvial indet.                 1601
lacustrine indet.              1240
                              ...  
platform/shelf-margin reef        3
lacustrine delta front            2
offshore ramp                     2
deep subtidal ramp                1
basin reef                        1
Name: count, Length: 70, dtype: int64

In [58]:
df['lithology1'].value_counts().head(30)

lithology1
not reported                       12032
sandstone                          11079
"siliciclastic"                     2569
mudstone                            2092
claystone                           1876
siltstone                           1775
"limestone"                          937
"shale"                              611
conglomerate                         605
marl                                 593
tar                                  592
lime mudstone                        270
breccia                              241
"carbonate"                          201
gravel                               168
wackestone                           144
peat                                 124
"mixed carbonate-siliciclastic"      110
phosphorite                          108
grainstone                            93
coal                                  82
dolomite                              76
chalk                                 41
tuff                                  40
ligni

In [63]:
not_reported = (df['lithology1'] == 'not reported').sum()
per_not_reported = not_reported/df.shape[0]
per_not_reported * 100

np.float64(31.816378877224526)

In [59]:
df["geological_group"].value_counts().head(30)

geological_group
Belly River      1134
Montana           651
Stormberg         330
Tremp             326
Glen Canyon       311
Agawam            303
Neuquén           285
Dakota            264
Wealden           263
Jehol             253
Enciso            242
Edmonton          238
Bauru             214
Trinity           169
Meriden           164
Chesapeake        160
Malargüe          157
Difunta           157
Rio do Peixe      155
Hayang            153
Rolling Downs     143
Oncala            129
Khorat            124
Puca              120
Kem Kem           112
Fort Munro         97
Chatham            95
Nanxiong           89
Wangshi            84
Purbeck            82
Name: count, dtype: int64

Geographic information difficult to parse right now. Lithology data has 31% of the the records listed as not reported. 

In [73]:
#Bias check
df["pres_mode"].value_counts()

pres_mode
body                                     25998
mold/impression,trace                     2350
body,anthropogenic                        1393
trace                                     1019
body,original phosphate                    870
                                         ...  
adpression,replaced with limonite            1
permineralized,original sporopollenin        1
cast,trace,replaced with phosphate           1
original sporopollenin                       1
adpression,coprolite                         1
Name: count, Length: 160, dtype: int64

In [74]:
df["preservation_quality"].value_counts()

preservation_quality
good         5793
medium       2830
excellent    1106
poor          901
variable      486
very poor      21
Name: count, dtype: int64

In [75]:
df["collection_coverage"].value_counts()

collection_coverage
some genera                                                      902
some macrofossils                                                684
all macrofossils                                                 234
some genera,species names                                        126
some macrofossils,some microfossils                              100
all microfossils                                                  79
some genera,some macrofossils                                     63
some microfossils                                                 46
some genera,some macrofossils,species names                       43
ichnofossils                                                      37
species names                                                     31
difficult macrofossils                                            31
all microfossils,some macrofossils                                28
some macrofossils,ichnofossils                                    21
some genera,so

In [76]:
df["collection_type"].value_counts()

collection_type
general faunal/floral    18539
taxonomic                13018
archaeological            2586
paleoecologic             1258
biostratigraphic          1215
taphonomic                 755
Name: count, dtype: int64

In [77]:
df["lagerstatten"].value_counts()

lagerstatten
concentrate     1349
conservation     147
Name: count, dtype: int64

In [78]:
df["research_group"].value_counts().head(20)

research_group
vertebrate                                          34165
taphonomy,vertebrate                                  414
PACED                                                 253
paleoentomology,vertebrate                            249
marine invertebrate                                   227
paleoentomology                                       220
marine invertebrate,vertebrate                        181
paleobotany                                            55
marine invertebrate,paleobotany,vertebrate             41
paleobotany,vertebrate                                 39
PACED,vertebrate                                       37
freshwater                                             36
freshwater,vertebrate                                  31
micropaleontology,paleobotany,vertebrate               17
micropaleontology,vertebrate                            4
decapod,marine invertebrate,vertebrate                  3
micropaleontology                                       3

In [65]:
#Data quality check
core_fields = [
    "occurrence_no",
    "collection_no",
    "identified_name",
    "accepted_name",
    "accepted_rank",
    "early_interval",
    "late_interval",
    "max_ma",
    "min_ma",
    "lat",
    "lng",
    "cc",
    "formation",
    "environment",
]

df[core_fields].isna().sum()

occurrence_no          0
collection_no          0
identified_name        0
accepted_name          0
accepted_rank          0
early_interval         0
late_interval      29133
max_ma                 0
min_ma                 0
lat                    0
lng                    0
cc                    17
formation          14328
environment          270
dtype: int64

In [66]:
(df[core_fields].isna().mean() * 100).round(2)

occurrence_no       0.00
collection_no       0.00
identified_name     0.00
accepted_name       0.00
accepted_rank       0.00
early_interval      0.00
late_interval      77.04
max_ma              0.00
min_ma              0.00
lat                 0.00
lng                 0.00
cc                  0.04
formation          37.89
environment         0.71
dtype: float64

In [67]:
df.duplicated().sum()

np.int64(0)

In [68]:
df["occurrence_no"].duplicated().sum()

np.int64(0)

In [71]:
(df['accepted_rank'] == 'species').sum()/df.shape[0]

np.float64(0.4578893090409075)

Dataset: 37,816 fossil occurrences from 14,351 fossil collections

Taxonomy: Dataset contains spieces, genus, family and higher-level identifications. 45% of entries are represted by speices name

Time: Occurences span approzimately 0-250 Ma

Geography: Records cover the entire globe and all fossil occurrences have geographic cooridnates

Geology: 1778 formations and 70 depositional environments are represented

Bias: Fossil quality found to be good or better for majority of occurrences and collection_type not as board as expected, most are in generic categories. 

Data Quality: Core analytical fields are mostly complete with a major expection with late_interval and formation having missing records. 